Crashlytics/Crashlytics/Rollouts/CrashlyticsRemoteConfigManager.swift

```diff
@@ -15,38 +15,56 @@
 import FirebaseRemoteConfigInterop
 import Foundation
 
-protocol CrashlyticsPersistentLog: NSObject {
-  func updateRolloutsStateToPersistence(rolloutAssignments: [RolloutAssignment])
+@objc(FIRCLSPersistenceLog)
+public protocol CrashlyticsPersistenceLog {
+  func updateRolloutsStateToPersistence(rollouts: Data, reportID: String)
 }
 
 @objc(FIRCLSRemoteConfigManager)
 public class CrashlyticsRemoteConfigManager: NSObject {
   public static let maxRolloutAssignments = 128
   public static let maxParameterValueLength = 256
 
+  private let lock = NSLock()
+  private var _rolloutAssignment: [RolloutAssignment] = []
+
   var remoteConfig: RemoteConfigInterop
-  @objc public private(set) var rolloutAssignment: [RolloutAssignment] = []
-  weak var persistenceDelegate: CrashlyticsPersistentLog?
+  var persistenceDelegate: CrashlyticsPersistenceLog
 
-  @objc public init(remoteConfig: RemoteConfigInterop) {
-    self.remoteConfig = remoteConfig
+  @objc public var rolloutAssignment: [RolloutAssignment] {
+    lock.lock()
+    defer { lock.unlock() }
+    let copy = _rolloutAssignment
+    return copy
   }
 
-  @objc public func updateRolloutsState(rolloutsState: RolloutsState) {
-    rolloutAssignment = normalizeRolloutAssignment(assignments: Array(rolloutsState.assignments))
+  @objc public init(remoteConfig: RemoteConfigInterop,
+                    persistenceDelegate: CrashlyticsPersistenceLog) {
+    self.remoteConfig = remoteConfig
+    self.persistenceDelegate = persistenceDelegate
   }
 
-  @objc public func getRolloutAssignmentsEncodedJson() -> String? {
-    let contentEncodedRolloutAssignments = rolloutAssignment.map { assignment in
-      EncodedRolloutAssignment(assignment: assignment)
+  @objc public func updateRolloutsState(rolloutsState: RolloutsState, reportID: String) {
+    lock.lock()
+    _rolloutAssignment = normalizeRolloutAssignment(assignments: Array(rolloutsState.assignments))
+    lock.unlock()
+
+    // writring to persistence
+    if let rolloutsData =
+      getRolloutsStateEncodedJsonData() {
+      persistenceDelegate.updateRolloutsStateToPersistence(
+        rollouts: rolloutsData,
```

Comment:
Curious why we're using NSData instead of NSString for this parameter?

Reply:
Oh because `[NSFileHandle writerData:]` need NSData as parameter. The other place for non-fatal I passed as NSString is because the helpers need input as NSString...